src/main/scala/goatrodeo/util/ArtifactWrapper.scala

```diff
@@ -6,22 +6,92 @@ import java.io.BufferedInputStream
 import java.io.FileInputStream
 import java.io.ByteArrayInputStream
 
-trait ArtifactWrapper {
+/** In OmniBOR, everything is seen as a byte stream.
+  *
+  * `ArtifactWrapper` provides a wrapper around artifacts... byte streams.
+  *
+  * `ArtifactWrapper` is somewhat modeled on the JVM `java.io.File` class so
+  * that it can represent both files on a filesystem with directories and
+  * virtual files contained in a Zip or TAR or other container file.
+  *
+  * Those byte streams may be in-memory or they may be on disk.
+  */
+sealed trait ArtifactWrapper {
+
+  /** Convert the Artifact to a stream of bytes. Note that this is mostly used
+    * for Hashing which is a block operation. No need for any buffering because
+    * all the operations will be pulling large blocks of bytes.
+    *
+    * @return
+    */
   def asStream(): InputStream
+
+  /** The name of the Artifact. This corresponds to the name of a `File` on disk
+    *
+    * @return
+    */
   def name(): String
+
+  /** The number of bytes in the artifact. This is
+    *
+    * @return
+    */
   def size(): Long
+
+  /** Corresponds to `File.isFile()`
+    *
+    * @return
+    */
   def isFile(): Boolean
+
+  /** Corresponds to `File.isDirectory()`
+    *
+    * @return
+    */
   def isDirectory(): Boolean
-  def isRealFile(): Boolean
-  // def asFile(): (File, Boolean)
-  def listFiles(): Vector[ArtifactWrapper]
+
+  /** Get all the `ArtifactWrappers` for the entities in a directory. This will
+    * only return something if the ArtifactWrapper is a wrapper around `File`
+    * and the entity being traversed is a real filesystem.
+    *
+    * For "container" file types (e.g., a TAR), see
+    * `FileWalker.streamForArchive`
+    *
+    * @return
+    */
+  def listFiles(): Iterator[ArtifactWrapper]
+
   def getCanonicalPath(): String
   def getParentDirectory(): File
   def delete(): Boolean
+
+  /** Does the entity exist. Corresponds to `File.exists`
+    *
+    * A `ByteWrapper` will always exist.
+    *
+    * @return
+    */
   def exists(): Boolean
+
+  lazy val suffix = ArtifactWrapper.suffix(name())
 }
 
-case class FileWrapper(f: File, deleteOnFinalize: Boolean)
+object ArtifactWrapper {
+
+  /** What's the suffix for the name
+    *
+    * @param name
+    *   the file name
+    * @return
+    *   the optional suffix to lower case
+    */
+  def suffix(name: String): Option[String] = {
```

Comment:
For files with multiple extensions, would the suffix ever need to include more than just the last one?

E.g. the suffix of `foo.tgz` is `tgz`, but the suffix of `bar.tar.gz` is `gz`.

Reply:
Yeah... this is a partial issue... mostly related to detection of zip/jar and iso... we need to see how file detection evolves.

My goal for ISO + VMDK + RPM + GEM + APK (Android) + ??? was to get a sense of the different factors related to the formats.

Once we have a bunch of less-than-optimal code that deals with the corner cases, exceptions, etc. then we can look at a broader solution.

This iteration is "make it work." The next few iterations with be "make it work and see what gets ugly."